v2/selective.go

```diff
@@ -0,0 +1,268 @@
+)
+
+// ErrSizeMismatch is returned when a written traversal realizes the written header size does not
+// match the actual number of car bytes written.
+var ErrSizeMismatch = fmt.Errorf("car-error-sizemismatch")
+
+// ErrOffsetImpossible is returned when specified paddings or offsets of either a wrapped carv1
+// or index cannot be satisfied based on the data being written.
+var ErrOffsetImpossible = fmt.Errorf("car-error-offsetimpossible")
+
+// MaxTraversalLinks changes the allowed number of links a selector traversal
+// can execute before failing.
+//
+// Note that setting this option may cause an error to be returned from selector
+// execution when building a SelectiveCar.
+func MaxTraversalLinks(MaxTraversalLinks uint64) Option {
+	return func(sco *Options) {
+		sco.MaxTraversalLinks = MaxTraversalLinks
+	}
+}
+
+// PrepareTraversal walks through the proposed dag traversal to learn it's total size in order to be able to
+// stream out a car to a writer in the expected traversal order in one go.
+func PrepareTraversal(ctx context.Context, ls *ipld.LinkSystem, root cid.Cid, selector ipld.Node, opts ...Option) (Writer, error) {
+	cls, cntr := loader.CountingLinkSystem(*ls)
+
+	c1h := carv1.CarHeader{Roots: []cid.Cid{root}, Version: 1}
+	headSize, err := carv1.HeaderSize(&c1h)
+	if err != nil {
+		return nil, err
+	}
+	if err := traverse(ctx, &cls, root, selector, ApplyOptions(opts...)); err != nil {
+		return nil, err
+	}
+	tc := traversalCar{
+		size:     headSize + cntr.Size(),
+		ctx:      ctx,
+		root:     root,
+		selector: selector,
+		ls:       ls,
+		opts:     opts,
+	}
+	return &tc, nil
+}
+
+// FileTraversal writes a car file matching a given root and selector to the
+// path at `destination` using one read of each block.
+func FileTraversal(ctx context.Context, ls *ipld.LinkSystem, root cid.Cid, selector ipld.Node, destination string, opts ...Option) error {
+	tc := traversalCar{
+		size:     0,
+		ctx:      ctx,
+		root:     root,
+		selector: selector,
+		ls:       ls,
+		opts:     opts,
+	}
+
+	fp, err := os.Create(destination)
+	if err != nil {
+		return err
+	}
+	defer fp.Close()
+
+	_, err = tc.WriteTo(fp)
+	if err != nil {
+		return err
+	}
+
+	// fix header size.
+	if _, err = fp.Seek(0, 0); err != nil {
+		return err
+	}
+
+	tc.size = uint64(tc.size)
+	if _, err = tc.WriteV2Header(fp); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// V1Traversal walks through the proposed dag traversal and writes a carv1 to the provided io.Writer
+func V1Traversal(ctx context.Context, ls *ipld.LinkSystem, root cid.Cid, selector ipld.Node, writer io.Writer, opts ...Option) (uint64, error) {
+	opts = append(opts, WithoutIndex())
+	tc := traversalCar{
+		size:     0,
+		ctx:      ctx,
+		root:     root,
+		selector: selector,
+		ls:       ls,
+		opts:     opts,
+	}
+
+	len, _, err := tc.WriteV1(writer)
+	return len, err
+}
+
+// Writer is an interface allowing writing a car with the data specified by a selector.
```

Comment:
I am curious what in this interface is selector-specific? 
* If it is meant to be selects-specific I'd recommend qualifying  that in the name since it's in the root of CARv2 library in case we add writers in the future.